### ISSUE TITLE
Using predicate functions on records pattern matching

### ISSUE THREAD

**Issue by lucasteles** [fsharp/fslang-suggestions#991]
# Using predicate functions on records pattern matching

My idea is to have a way to apply predicate functions to records on match 

Something like this.

```fsharp
let isEven n = n % 2 = 0
let hasMinLength size (s: string) = s.Length > size
type Foo = { Name: string; Bar: int; Baz: int }

let assertFoo = 
   function
   | { Bar: isEven
       Name: hasMinLength 3
       Baz:  > 10 } -> true
   | _ -> false

```

## Pros and Cons 

Pros: 
- Aids code readability in specialized cases.
- Can made better unit test asserts
- Reduce boilerplate

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
* [x] I or my company would be willing to help implement and/or test this

## For Readers

If you would like to see this issue implemented, please click the :+1: emoji on this issue. These counts are used to generally order the suggestions by engagement.

**Comment by kerams** [comment]
> Reduce boilerplate

Is it really better than this
```fsharp
let assertFoo x = 
    isEven x.Bar &&
    hasMinLength 3 x.Name &&
    x.Baz > 10
```

or do you have a better example?

**Comment by lucasteles** [comment]
@kerams for this example I don't think so


Cases that I believe it could be better is for more complex cases that not return a boolean and/or nested members, or unit test assertions

this case maybe worth it, idyt?
```fs
let foo = 
   function
   | { Bar: isEven;  Name: hasMinLength 3;  Baz:  > 10 } -> printfn "match a"
   | { Bar: isEven >> not;  Name: hasMinLength 5;  Baz=42 } -> printfn "match b"
   | { Bar: isEven;  OtherProp = {  OtherValue:  isEven  }} -> printfn "match c"
   | _ -> printfn "match d"
```

For unit tests is common to have a partial assertion, if we had a way to apply this `shape` as a boolean func, it could be very expressive

```fs
someFn data
|> match
      {  A="Expected value a"
          B="Expected value b"
          Id: isAGuid  // i dont care about the Id here. it only have to be a guid
        }
```

this maybe makes the assertion more clear, because is closer to the original data shape.

**Comment by Happypig375** [comment]
It's called active patterns.
```fs
let​ ​(|IsEven|_|)​ n ​=​ if n % ​2​ ​=​ ​0​ then Some IsEven else None
​let​  (|HasMinLength|_|)​ size ​(​s​:​ ​string​)​ ​=​ if s.Length ​>​ size then Some HasMinLength else None
let (|GreaterThan|_|) n x = if x > n then Some GreaterThan else None
​type​ ​Foo​ ​=​ ​{​ Name​:​ ​string​;​ Bar​:​ ​int​;​ Baz​:​ ​int ​}​

​let​ ​assertFoo​ ​=​ 
   ​function​
   ​|​ ​{​ Bar = IsEven​
       Name = HasMinLength 3​
       Baz​:​ ​GreaterThan​ 10 ​}​ ​->​ ​true​
   ​|​ ​_​ ​->​ ​false​
```

**Comment by Tarmil** [comment]
I agree with @Happypig375. And if you don't want to define all your predicates as active patterns, you can just make one:

```fsharp
let (|Matches|_|) f x =
    if f x then Some x else None

let assertFoo = function
    | { Bar = Matches isEven _
        Name = Matches (hasMinLength 3) _
        Baz = Matches (greaterThan 10) _ } -> true
    | _ -> false
```

**Comment by cartermp** [comment]
Yep, this is unlikely to happen since it's squarely within the Active Pattern space.

**Comment by lucasteles** [comment]
@Tarmil perfect!
